### PR TITLE
KAFKA-16592: Add a new constructor which invokes the existing constructor with default value for alternativeString

### PR DESCRIPTION
**Pull request by vamossagar12**
Description can be found in [KAFKA-16592](https://issues.apache.org/jira/browse/KAFKA-16592)

### CODE REVIEW

**Comment by frankvicky** [comment]
@chia7712  Could you please assist me with reviewing this pull request ? 😃

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 1]
```diff
@@ -1256,6 +1256,15 @@ public static class ConfigKey {
         public final boolean internalConfig;
         public final String alternativeString;
 
+        public ConfigKey(String name, Type type, Object defaultValue, Validator validator,
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:1260]: Could you please add comments to say that we keep this method due to backward compatibility?

**Review by chia7712 — COMMENTED** [review]
@vamossagar12 nice finding!

Users are allowed to create `ConfigKey` by either "define" or "constructor". It seems to me that keeping both way can obstruct us from developing related code. #13909 is a example. Maybe we should file a KIP to deprecate one of them....

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 1 above)
Comment (vamossagar12) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java:1260]: Done.

**Comment by vamossagar12** [comment]
Thanks @chia7712 , I am thinking we can make the constructors as private and let the users use only `define` method. Even in the AK codebase, I don't see the constructor being used widely and all declarations are being done using `define`. I can file a ticket for this if we agree to this.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 2]
```diff
@@ -1256,6 +1256,16 @@ public static class ConfigKey {
...
+                         Importance importance, String documentation, String group,
+                         int orderInGroup, Width width, String displayName,
+                         List<String> dependents, Recommender recommender,
+                         boolean internalConfig) {
+            this(name, type, defaultValue, validator, importance, documentation, group, orderInGroup, width, displayName,
+                dependents, recommender, internalConfig, null);
+        }
+
         public ConfigKey(String name, Type type, Object defaultValue, Validator validator,
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: As #13909 is not in any release, we can change this constructor from `public` to `private`

**Review by chia7712 — COMMENTED** [review]
@vamossagar12 thanks for updated PR.

> I am thinking we can make the constructors as private and let the users use only define method. Even in the AK codebase, I don't see the constructor being used widely and all declarations are being done using define. I can file a ticket for this if we agree to this.

According to our BC, I guess we need a KIP to deprecate that method and add comments to encourage users to use `define`. Finally, we can remove the constructor after deprecation cycle.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 2 above)
Comment (vamossagar12) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Done

**Comment by vamossagar12** [comment]
Thanks @chia7712 , for now I have made the constructor added as part of https://github.com/apache/kafka/pull/13909 as private. 
Regarding  

> I guess we need a KIP to deprecate that method and add comments to encourage users to use

it would mostly be making the other constructor firstly deprecated and finally private. I don't think we can remove the Constructor because `define` eventually depends upon it. I will create a new ticket for the KIP to track it.

**Comment by chia7712** [comment]
retrigger QA due to incomplete build. I will merge it after QA get completed

**Comment by vamossagar12** [comment]
@chia7712 , I created this ticket: [KAFKA-16604](https://issues.apache.org/jira/browse/KAFKA-16604) for tracking the KIP

**Comment by vamossagar12** [comment]
The build got aborted again 🤔

**Comment by chia7712** [comment]
@vamossagar12 could you please rebase code? I guess CI is too busy to run tests :(

**Comment by chia7712** [comment]
I don't see any related failure, so I'm going to merge it to trunk and branch 3.7

**Pull request merged**